src/main/clojure/clara/rules/platform.cljc

```diff
@@ -7,10 +7,25 @@
   [^String description]
   (throw #?(:clj (IllegalArgumentException. description) :cljs (js/Error. description))))
 
+;; This class wraps Clojure objects to ensure Clojure's equality and hash
+;; semantics are visible to Java code. This allows these Clojure objects
+;; to be safely used in things like Java Sets or Maps.
+;; This class also accepts and stores the hash code, since it almost always
+;; will be used once and generally more than once.
+#?(:clj
+   (deftype JavaEqualityWrapper [wrapped ^int hash-code]
+
+     Object
+     (equals [this other]
+       (= wrapped (.wrapped ^JavaEqualityWrapper other)))
```

Comment:
You are assuming the `other` is always going to be a `JavaEqualityWrapper`, otherwise there could be cast exceptions. I believe that is fine though, since this is only intended to be used/compared in collections that contain only these `JavaEqualityWrapper` wrapper types.

To confirm: this is your reasoning here, correct?

Reply:
That is the only use case I had in mind, but it's a good idea to add an instance? check here. They're quite cheap and makes this class more general.